src/kaleidoscope_internal/plugin_hooks.h

```diff
@@ -90,118 +90,56 @@ namespace kaleidoscope_internal {
-         trait class. */                                                  __NL__ \
-      typedef typename ReturnTypeTraits<                                  __NL__ \
-                 decltype(&kaleidoscope::Plugin::PLUGIN_METHOD)           __NL__ \
-              >::Type  ReturnType;                                        __NL__ \
+      static bool shouldAbortOnConsumedEvent() {                          __NL__ \
+        return SHOULD_ABORT_ON_CONSUMED_EVENT;                            __NL__ \
+      }                                                                   __NL__ \
                                                                           __NL__ \
       template<typename Plugin__, typename... Args__>                     __NL__ \
-      static ReturnType call(Plugin__ &plugin, Args__&&... hook_args) {   __NL__ \
-         /* Assert that hook methods in derived plugins have              __NL__ \
-            signatures that match the base class. */                      __NL__ \
+      static kaleidoscope::Plugin::Result call(Plugin__ &plugin,          __NL__ \
+                                               Args__&&... hook_args) {   __NL__ \
          _VALIDATE_HOOK_METHOD_SIGNATURE(PLUGIN_METHOD, Plugin__)         __NL__ \
                                                                           __NL__ \
          return plugin.PLUGIN_METHOD(hook_args...);                       __NL__ \
       }                                                                   __NL__ \
-   };
-
-// Predicate classes that decides if the hook-plugin-loop should abort early.
-
-struct AbortIfFalse {
-  static bool eval(bool val) {
-    return !val;
-  }
 };
 
-struct AlwaysContinue {
-  constexpr static bool eval(bool val) {
-    return false;
-  }
-};
-
-_PLUGIN_METHOD(onSetup, AlwaysContinue)
-_PLUGIN_METHOD(beforeEachCycle, AlwaysContinue)
-_PLUGIN_METHOD(onEvent, AbortIfFalse)
-_PLUGIN_METHOD(beforeReportingState, AlwaysContinue)
-_PLUGIN_METHOD(afterEachCycle, AlwaysContinue)
+_PLUGIN_METHOD(onSetup, false)
+_PLUGIN_METHOD(beforeEachCycle, false)
+_PLUGIN_METHOD(onEvent, true)
+_PLUGIN_METHOD(beforeReportingState, false)
+_PLUGIN_METHOD(afterEachCycle, false)
 
 // The following _CALL macros are used by DEFINE_HOOKPOINT in
 // conjunction with the MAP macro to call the hook method in each listed plugin.
 
 #define _CALL_HOOK_FOR_PLUGIN(PLUGIN)                                       \
    result = PluginMethod__::call(PLUGIN, hook_args...);              __NL__ \
                                                                      __NL__ \
-   if (ShouldAbort::eval(result)) {                                  __NL__ \
+   if (PluginMethod__::shouldAbortOnConsumedEvent() &&               __NL__ \
+       result == kaleidoscope::Plugin::Result::EVENT_CONSUMED) {     __NL__ \
       return result;                                                 __NL__ \
    }                                                                 __NL__
 
```

Comment:
Currently the `Result` option `ERROR` is completely ignored and thus problematic as it might be returned intentionally. I would at least add an assertion to the end of `_CALL_HOOK_FOR_PLUGIN`
```cpp
assert(result != HandlerResult::Error);
```

Mind you that `-DNDEBUG` is probably still missing from the kaleidoscope_builder's default compile flags.

Reply:
It is ignored now on purpose: in case of an error, we want the rest of the chain to process it still. At least until we figure out a way to signal that error to the end-user.

Good point about assert & NDEBUG, though.

Reply:
I am afraid, I am still not convinced.

> until we figure out a way to signal that error to the end-user.

I suppose the resources of the firmware are too limited to even think about any kind of runtime error handling. There is no point in passing error flags around. It will always require proper debugging to fix those errors and that's where assertions come in handy. 

> It is ignored now on purpose: in case of an error, we want the rest of the chain to process it still.

Even if you stick with the `ERROR` return value, an assertion does ignore it in release builds. So it definitely does what you want. Only in debug builds the assertion will report it - until you come up with a better idea. Even though it is not yet possible to generate debug builds with kaleidsocope_builder, it definitely makes sense to add as many assertions as necessary right now, while developing. As soon as there will be debug builds - and I hope they will be available one day - assertions will become useful.

Every decent FOSS C++ project uses assertions. Why should we go without?

Assertions should be added in all critical places, during development of new code and when fixing errors in order to prevent regressions. During build-upload-test cycles assertions can help to track down errors faster. And when a host-based debugger/simulator is used, assertions help to stop the program at the right place.